util/logtools/arg_parser.ts

```diff
@@ -0,0 +1,87 @@
+import argparse from 'argparse';
+
+type TimelineArgs = {
+  [index: string]: string | number | boolean | undefined;
+  'file'?: string;
+  'force'?: boolean;
+  'search_fights'?: number;
+  'search_zones'?: number;
+  'fight_regex'?: string;
+  'zone_regex'?: string;
+  'adjust'?: number;
+};
+
+class LogUtilArgParse {
+  parser = new argparse.ArgumentParser({
+    addHelp: true,
+  });
+  // At least one non-file argument must be selected from this group.
+  fileGroup = this.parser.addArgumentGroup({
+    title: 'File Args',
+    description: 'A file and/or a timeline must be selected before performing operations',
+  });
+  // Exactly one argument should be selected from this group.
+  requiredGroup = this.parser.addMutuallyExclusiveGroup({ required: true });
+  args: TimelineArgs;
+
+  constructor() {
+    this.fileGroup.addArgument(['-f', '--file'], {
+      help: 'Network log file to analyze',
+    });
+    this.fileGroup.addArgument(['-t', '--timeline'], {
+      help: 'The filename of the timeline to test against, e.g. ultima_weapon_ultimate',
+    });
+    this.parser.addArgument(['--force'], {
+      nargs: 0,
+      help: 'Overwrite files when exporting',
+    });
+    this.requiredGroup.addArgument(['-lf', '--search-fights'], {
+      nargs: '?',
+      constant: -1,
+      type: 'int',
+      help: 'Fight in log to export, e.g. \'1\'. ' +
+        'If no number is specified, returns a list of fights.',
+    });
+    this.requiredGroup.addArgument(['-lz', '--search-zones'], {
+      nargs: '?',
+      constant: -1,
+      type: 'int',
+      help: 'Zone in log to export, e.g. \'1\'. ' +
+        'If no number is specified, returns a list of zones.',
+    });
+    this.requiredGroup.addArgument(['-fr', '--fight-regex'], {
+      nargs: '1',
+      type: 'string',
+      help: 'Export all fights that match this regex',
+    });
+    this.requiredGroup.addArgument(['-zr', '--zone-regex'], {
+      nargs: '1',
+      type: 'string',
+      help: 'Export all zones that match this regex',
+    });
+    this.requiredGroup.addArgument(['-a', '--adjust'], {
+      nargs: '?',
+      constant: 0,
+      type: 'float',
+      help: 'Adjust all entries in a timeline file by this amount',
+    });
+    this.args = this.validate(this.parser.parseArgs() as TimelineArgs);
+  }
+
+  validate(args: TimelineArgs): TimelineArgs {
```

Comment:
> I don't like the games I had to play with argsObject, but there's not much that can be done when the library itself returns any from a core method. It works, at least.

For what it's worth, you could make this a [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards), e.g.

```typescript
const args = this.parser.parseArgs();
if (isTimelineArgs(args))
  this.args = args; // args is implicitly TimelineArgs inside this conditional due to type guard
if (!this.args)
  process.exit(-1);

// ...and then elsewhere...

isTimelineArgs(args: unknown): args is TimelineArgs {
  // TODO: first make sure args is of type object and not an array

  const validArgs: Required<TimelineArgs> {
    file: '',
    // fill out rest of args here
  };
  for (const [key, value] of Object.entries(validArgs)) {
    if (key in args && typeof args[key] !== typeof value)
      return false;
  }
  return true;
}

Reply:
I'll poke at this once I finish up with the timeline conversions. This suggestion would definitely be less potentially fallible than what I did.